myia/ir.py

```diff
@@ -1,7 +1,222 @@
+        self.parameters: List[Parameter] = []
+        self.return_: Optional[Return] = None
+        self.debug: Dict = {}
 
 
 class Node:
-    """A node in the graph-based ANF IR."""
+    """A node in the graph-based ANF IR.
+
+    There are four types of nodes: Function applications; parameters; return
+    values; and constants such as numbers and functions.
+
+    Attributes:
+        inputs: If the node is a function application, the first node input is
+            the function to apply, followed by the arguments. These are use-def
+            edges.
+        value: The value of this node, if it is a constant. Parameters have the
+            special value `PARAMETER`.
+        graph: The function definition graph that this node belongs to for
+            values and parameters.
+        uses: A set of tuples with the nodes that use this node alongside with
+            the index. These def-use edges are the reverse of the `inputs`
+            attribute, creating a doubly linked graph structure.
+        debug: A dictionary with debug information about this node e.g. a
+            human-readable name and the Python source code.
+
+    """
+
+    def __init__(self, inputs: List['Node'], value: Any,
+                 graph: Optional[Graph]) -> None:
+        """Construct a node."""
+        self.inputs = Inputs(self, inputs)
+        self.value = value
+        self.graph = graph
+        self.uses: Set[Tuple[Node, int]] = set()
+        self.debug: Dict = {}
+
+
+class Inputs(MutableSequence[Node]):
+    """Container data structure for node inputs.
+
+    This mutable sequence data structure can be used to keep track of a node's
+    inputs. Any insertion or deletion of edges will be reflected in the inputs'
+    `uses` attribute.
+
+    """
+
+    def __init__(self, node: Node, initlist: Iterable[Node] = None) -> None:
+        """Construct the inputs container for a node.
+
+        Args:
+            node: The node of which the inputs are stored.
+            initlist: A sequence of nodes to initialize the container with.
+
+        """
+        self.node = node
+        self.data: List[Node] = []
+        if initlist is not None:
+            self.extend(initlist)
+
+    @overload
+    def __getitem__(self, index: int) -> Node:
+        pass
+
+    @overload  # noqa: F811
+    def __getitem__(self, s: slice) -> Sequence[Node]:
+        pass
+
+    def __getitem__(self, index):  # noqa: F811
+        """Get an input by its index."""
+        return self.data.__getitem__(index)
+
+    @overload
+    def __setitem__(self, index: int, value: Node) -> None:
+        pass
+
+    @overload  # noqa: F811
+    def __setitem__(self, index: slice, value: Iterable[Node]) -> None:
+        pass
+
+    def __setitem__(self, index, value):  # noqa: F811
+        """Replace an input with another."""
+        if isinstance(index, slice):
+            for i, v in zip(range(*index.indices(len(self))), value):
+                self[i] = v
```

Comment:
This doesn't have the same semantics as python when you assign a sequence that has a different length than the indices.

Reply:
You mean that Python crashes in that case but this doesn't?

Reply:
He means that assigning to a slice in Python is a splicing operation, e.g.

```python
a = ['a', 'b', 'c', 'd']
a[1:3] = range(10)
a  # ['a', 1, 2, 3, 4, 5, 6, 7, 8, 9, 'd']
```

Your code would, I suppose, produce `['a', 1, 2, 'd']` instead.

Reply:
Ah, I forgot about that. (I tried `a[1:3:2] = range(10)`, which crashes; only continuous slices can be assigned differently sized iterables.)